dlib/dnn/cudnn_dlibapi.cpp

```diff
@@ -118,6 +118,17 @@ namespace dlib
             return c.get_handle();
         }
 
+        static std::shared_ptr<resizable_buffer> device_global_buffer()
+        {
+            thread_local std::weak_ptr<resizable_buffer> weak_buff;
```

Comment:
You need to add a call to cudaGetDevice() here (like on line 97 above) to make sure you are returning a resizable_buffer that's active on the currently selected cuda device.  Otherwise there is still the problem of someone switching the active thread's cuda device and then trying to use dlib, which would cause a crash or other huge problem with this code as it is.